netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -446,6 +448,25 @@ public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
     }
 
     /**
+     * Do not rely on channel handlers to propagate exceptions to us.
+     * {@link NettyClientHandler} is an example of a class that does not propagate exceptions.
+     * Add a listener to the connect future directly and do appropriate error handling.
+     */
+    @Override
+    public void connect(final ChannelHandlerContext ctx, SocketAddress remoteAddress,
+        SocketAddress localAddress, ChannelPromise promise) throws Exception {
+      promise.addListener(new ChannelFutureListener() {
```

Comment:
Add the promise after calling `connect()`. Always add the promise afterward for sane ordering.